### PR TITLE
Fix function calling, more verbose logging, OpenAI upgrade

### PR DESCRIPTION
**Pull request by undernagruzez**
### Motivation and Context
Langchain release a new major upgrade. The version in module programming llm was far behind and had to be upgraded. This PR addresses this and also adds a bit better logging.

### Description

What was changed:
1. The way llm chains are built - was updated to use the new features of langchain.
2. The way LLM models are initialized - because of stuff removed from the long chain.
3. A flag to support more verbose logging.
4. Repeated LLM invocations - in case of failures, additional invocations will be issued, up to 3 times.

### Steps for Testing

Just deploy it to the test server and request feedback and suggestions from Athena and Artemis.

### CODE REVIEW

**Comment by FelixTJDietrich** [comment]
What is up with this PR? Can you update it? Also the PR description is missing. 

If it is not ready can you please put it in draft until it is ready. Also try to find reviewers for the PR once it is ready.

**Comment by undernagruzez** [comment]
> What is up with this PR? Can you update it? Also the PR description is missing.
> 
> If it is not ready can you please put it in draft until it is ready. Also try to find reviewers for the PR once it is ready.

This PR is ready to review. I updated the description.

File: module_programming_llm/module_programming_llm/helpers/models/openai.py [hunk 1]
```diff
@@ -142,46 +142,62 @@ def _openai_client(use_azure_api: bool, is_preference: bool):
...
-                if model_name in model_aliases:
-                    model_name = model_aliases[model_name]
-                if model_name in openai_models["chat_completion"]:
+                if deployment.capabilities["chat_completion"]:
                     available_deployments["chat_completion"][deployment.id] = deployment
-                elif model_name in openai_models["completion"]:
-                    available_deployments["completion"][deployment.id] = deployment
-                elif model_name in openai_models["fine_tuneing"]:
-                    available_deployments["fine_tuneing"][deployment.id] = deployment
+                # if deployment.capabilities["completion"]:
```
Comment (FelixTJDietrich) [inline module_programming_llm/module_programming_llm/helpers/models/openai.py]: Remove commented out code?

File: module_programming_llm/module_programming_llm/helpers/models/openai.py [hunk 2]
```diff
@@ -207,39 +222,14 @@ def _get_available_models(openai_models: Dict[str, List[str]],
...
-        "davinci",
-        "curie",
-        "babbage",
-        "ada",
-    ]
-}
-available_deployments = _get_available_deployments(openai_models, _model_aliases)
-available_models = _get_available_models(openai_models, available_deployments)
+available_deployments = _get_available_deployments()
+available_models = _get_available_models(available_deployments)
```
Comment (FelixTJDietrich) [inline module_programming_llm/module_programming_llm/helpers/models/openai.py]: We are only supporting models that are deployed on Azure OpenAI? What about OpenAI models that are not deployed on Azure? We also want to be able to use them

File: athena/athena/models/model.py [hunk 3]
```diff
@@ -1,7 +1,8 @@
 import importlib
 
-from pydantic import BaseModel
-from sqlalchemy import Column, String, UniqueConstraint
+from pydantic import BaseModel, AnyUrl
```
Comment (FelixTJDietrich) [inline athena/athena/models/model.py]: There are some unused imports

**Review by FelixTJDietrich — CHANGES_REQUESTED** [review]

**Comment by FelixTJDietrich** [comment]
On the test server the programming LLM and the text LLM modules are not working.

**Comment by undernagruzez** [comment]
Handled here: https://github.com/ls1intum/Athena/pull/342

**Pull request closed without merging**